### PR TITLE
DOC: 2.0 release highlights and compat notes changes

### PR DESCRIPTION
**Pull request by rgommers**
This is a draft of the narrative section of the 2.0.0 release notes, with highlights and main messages. I think it needs to be significantly longer than that of a normal feature release, and give the reader an overview of what's in this release as well as links to other key documentation like the migration guide and the page about packaging for downstream libraries.

For the highlights I chose six top-level categories and then 2-4 items for each of those categories. Which are:

- New features
- Performance improvements
- Python API improvements
- C API improvements
- Improved behavior
- Documentation

A *lot* of work has gone into this release, so I'm sure this isn't 100% complete or representative yet. A few questions to promt feedback on this draft:

- Are the above six categories the right ones?
- Are they in the right order? (note: I had it differently first, but noticed that starting with new features and performance is more concrete for the average user, and hence probably more appealing)
- What is missing? And what did I include that you think should not be there?

This PR also takes care of some leftover tasks for release notes and migration guide.

### CODE REVIEW

**Comment by rgommers** [comment]
Let me ping a few people for specific categories:

- Documentation: @Mukulikaa, @rossbar, @melissawm 
- Static typing: @BvB93 (no major changes I think, but want to make sure)
- Performance: @seiko2plus, @r-devulap, @Mousius 
- Platform support: @andyfaff (no major changes here either I think; musllinux went into 1.26.x) 
- Any user-facing ufunc/gufunc features worth calling out that I missed? @seberg @mhvk

File: doc/source/release/2.0.0-notes.rst [hunk 1]
```diff
@@ -2,6 +2,82 @@
...
+- New features:
+
+  - A new variable-length string dtype, `~numpy.dtypes.StringDType` and a new
+    `numpy.strings` namespace with performant ufuncs ...
+  - Support for ``float32`` and ``longdouble`` in all `numpy.fft` functions,
+  - Support for the array API standard in the main ``numpy`` namespace
+
+- Performance improvements:
+
+  - Sorting functions (`sort`, `argsort`, `partition`)
```
Comment (r-devulap) [inline doc/source/release/2.0.0-notes.rst:21]:
```suggestion
  - Sorting functions (`sort`, `argsort`, `partition`, `argpartition`)
```

File: doc/source/release/2.0.0-notes.rst [hunk 2]
```diff
@@ -318,6 +394,13 @@ before.
 
 (`gh-25336 <https://github.com/numpy/numpy/pull/25336>`__)
 
+Minor changes in behavior of sorting functions
+----------------------------------------------
+
+Due to algorithmic changes and use of SIMD code, sorting functions with methods
+that aren't stable may return slightly different results in 2.0.0 compared to
+1.26.x. This includes the default method of `~numpy.sort` and `~numpy.argsort`.
```
Comment (r-devulap) [inline doc/source/release/2.0.0-notes.rst:402]:
```suggestion
1.26.x. This includes the default method of `~numpy.argsort` and `~numpy.argpartition`.
```

The unstableness in quicksort shows up in the `arg` methods only.

File: doc/source/release/2.0.0-notes.rst [hunk 3]
```diff
@@ -2,6 +2,82 @@
...
+
+  - A new variable-length string dtype, `~numpy.dtypes.StringDType` and a new
+    `numpy.strings` namespace with performant ufuncs ...
+  - Support for ``float32`` and ``longdouble`` in all `numpy.fft` functions,
+  - Support for the array API standard in the main ``numpy`` namespace
+
+- Performance improvements:
+
+  - Sorting functions (`sort`, `argsort`, `partition`)
+    have been accelerated through use of SIMD kernels and Highway, and may
```
Comment (r-devulap) [inline doc/source/release/2.0.0-notes.rst:22]:
```suggestion
    have been accelerated through the use of Intel x86-simd-sort and Google Highway libraries, and may
```

**Review by r-devulap — COMMENTED** [review]
Performance section looks good to me. Just some minor edits, if it seems appropriate.

File: doc/source/release/2.0.0-notes.rst [hunk 4]
```diff
@@ -2,6 +2,82 @@
 NumPy 2.0.0 Release Notes
 =========================
 
+NumPy 2.0.0 is the first major release since 2006. It is the result of X months
+of development since the last feature release by Y contributors, and contains a
+large amount of exciting new features as well as a large amount of changes to
+both the Python and C APIs.
+
```
Comment (rgommers) [inline doc/source/release/2.0.0-notes.rst:9]: Comment by @charris: start with a list of links to other docs that are relevant (migration guide, etc)

File: doc/source/release/2.0.0-notes.rst (hunk 1 above)
Comment (rgommers) [reply doc/source/release/2.0.0-notes.rst]: Thanks @r-devulap. Useful comments, all incorporated now.

**Comment by mattip** [comment]
[here](https://output.circle-artifacts.com/output/job/125c96cf-ccef-45a6-8213-2d8b9fec268b/artifacts/0/doc/build/html/release/2.0.0-notes.html) is the rendered document, if that is easier for people to read.

File: doc/source/numpy_2_0_migration_guide.rst [hunk 5]
```diff
@@ -8,6 +8,25 @@ This document contains a set of instructions on how to update your code to
 work with Numpy 2.0.
 
 
+Ruff plugin
+===========
+
+Many of the changes covered in the 2.0 release notes and in this migration
+guide can be automatically adapted to in downstream code with a dedicated
+`Ruff <https://docs.astral.sh/ruff/>`__ rule, namely rule
+`NPY201 <https://docs.astral.sh/ruff/rules/numpy2-deprecation/>`_.
```
Comment (lysnikolaou) [inline doc/source/numpy_2_0_migration_guide.rst:17]: Probably better to use double underscore in both links?

```suggestion
`NPY201 <https://docs.astral.sh/ruff/rules/numpy2-deprecation/>`__.
```

File: doc/source/numpy_2_0_migration_guide.rst [hunk 6]
```diff
@@ -383,3 +388,23 @@ Note about pickled files
...
+
+The :ref:`copy keyword behavior changes <copy-keyword-changes-2.0>` in
+`~numpy.asarray`, `~numpy.array` and `ndarray.__array__
+<numpy.ndarray.__array__>` may require these changes:
+
+1. Code using ``np.array(..., copy=False)`` can in most cases be changed to
+   ``np.asarray(...)``. Older code tended to use ``np.array`` like this because
+   it had less overhead than the default ``np.asarray`` copy-if-needed
+   behavior. This is no longer true, and ``np.asarray`` is the preferred function.
+2. For code that explicitly needs to pass ``None/False`` meaning "copy if
```
Comment (lysnikolaou) [inline doc/source/numpy_2_0_migration_guide.rst:404]:
```suggestion
2. For code that explicitly needs to pass ``None``/``False`` meaning "copy if
```

File: doc/source/release/2.0.0-notes.rst [hunk 7]
```diff
@@ -4,6 +4,93 @@
...
+    namespace decreased by ~10% and in ``numpy.lib`` by ~80%,
+  - :ref:`Canonical dtype names <canonical-python-and-c-types>` and a new
+    `~numpy.isdtype` introspection function,
+
+- C API improvements:
+
+  - A new public C API for creating custom dtypes
+  - Many outdated functions and macros removed, and private internals hidden to
+    ease future extensibility,
+  - New, easier to use, initialization functions,
```
Comment (lysnikolaou) [inline doc/source/release/2.0.0-notes.rst:60]: Is it standard to include trailing commas in list items?

File: doc/source/release/2.0.0-notes.rst [hunk 8]
```diff
@@ -4,6 +4,93 @@
...
+API and behavior improvements and better future extensibility. This price is:
+
+1. Backwards compatibility - clear errors in most cases; see the Deprecation
+   and Compatibility sections below, and the :ref:`numpy-2-migration-guide`
+   (note: there is a ``ruff`` mode to auto-fix many things in Python code)
+2. ABI change - impact on downstream libraries and binaries. See
+   :ref:`numpy-2-abi-handling`.
+   TODO: main messages.
+
+
```
Comment (lysnikolaou) [inline doc/source/release/2.0.0-notes.rst:90]: Delete second empty line.

```suggestion

```

File: doc/source/release/2.0.0-notes.rst [hunk 9]
```diff
@@ -4,6 +4,93 @@
...
+The Python versions supported by this release are 3.9-3.12.
+
+
```
Comment (lysnikolaou) [inline doc/source/release/2.0.0-notes.rst:93]: Delete third empty line
```suggestion

```

File: doc/source/release/2.0.0-notes.rst [hunk 10]
```diff
@@ -4,6 +4,93 @@
...
+  - Support for the array API standard in the main ``numpy`` namespace
+
+- Performance improvements:
+
+  - Sorting functions (`sort`, `argsort`, `partition`, `argpartition`)
+    have been accelerated through the use of the Intel x86-simd-sort and Google
+    Highway libraries, and may see large (hardware-specific) speedups,
+  - macOS Accelerate support and binary wheels for macOS >=14, with significant
+    performance improvements for linear algebra operations on macOS
+  - `numpy.char` fixed-length string dtypes have been accelerated, ...
```
Comment (lysnikolaou) [inline doc/source/release/2.0.0-notes.rst:39]:
```suggestion
  - `numpy.char` string operations have been accelerated by implementing
    ufuncs that also support `~numpy.dtypes.StringDType` in addition to the
    the fixed-length string dtypes.
```

**Review by lysnikolaou — COMMENTED** [review]
Not sure if this is ready for review yet, but here's a first round.

**Comment by mattip** [comment]
Please sneak this snippet in, it was left out of #25914:
```
Removed ambiguity when broadcasting in ``np.solve``
---------------------------------------------------
The broadcasting rules for ``np.solve(a, b)`` were ambiguous when ``b`` had 1
fewer dimensions than ``a``. This has been resolved in a backward-incompatible
way and is now compliant with the Array API. The old behaviour can be
reconstructed by using ``np.solve(a, b[..., None])[..., 0]``.

(`gh-25914 <https://github.com/numpy/numpy/pull/25914>`__) 
```

cc @asmeurer

File: doc/source/release/2.0.0-notes.rst [hunk 11]
```diff
@@ -4,6 +4,93 @@
...
+- C API improvements:
+
+  - A new public C API for creating custom dtypes
+  - Many outdated functions and macros removed, and private internals hidden to
+    ease future extensibility,
+  - New, easier to use, initialization functions,
+
+- Improved behavior:
+
+  - Improvements to type promotion behavior (NEP 50) ...
```
Comment (seberg) [inline doc/source/release/2.0.0-notes.rst:64]:
```suggestion
  - Improvements to type promotion behavior was changed by adopting `NEP 50<NEP50>`_.
    This makes fixes many user surprises about promotions which previously often
    dependet on the values.  Please see the NEP and the :ref:`numpy-2-migration-guide`
    for details as this change e.g. can lead to lower precision results.
```

File: doc/source/release/2.0.0-notes.rst [hunk 12]
```diff
@@ -4,6 +4,93 @@
 NumPy 2.0.0 Release Notes
 =========================
 
+NumPy 2.0.0 is the first major release since 2006. It is the result of X months
+of development since the last feature release by Y contributors, and contains a
+large amount of exciting new features as well as a large amount of changes to
+both the Python and C APIs.
```
Comment (seberg) [inline doc/source/release/2.0.0-notes.rst:10]: I am wondering if we can expand this section in some free-form to include:

Beyond general improvements and new features the major changes for Python users are Python API changes which for the most part are removals and cleanups but may require code adaptations (NEP 56) and the changes in promotion (NEP 50) which fix many user surprises but can lead to different precision results.

I.e. these are the most relevant (to end users) changes which couldn't happen in a 1.x release.  NEP 50 feels a bit lost below considering that it has quite a bit of potential for breaking things.
(The migration guide makes that clear I think, so I guess if you don't like listing anything here, you could also just stress that as a major release 2.0 had some changes which could affect users more, please see there there...)

**Review by seberg — COMMENTED** [review]
Not sure how, so fine.  One thing that I wonder about is that e.g. NEP 50 and the Python removals is what users will actually care about most in terms of compatibility and in terms of being the largest changes and improvements in many ways.

I would be happy to iterate, thus the thought of adding a note.

File: doc/source/release/2.0.0-notes.rst (hunk 7 above)
Comment (rgommers) [reply doc/source/release/2.0.0-notes.rst:73]: It's common; there's no style guide with this level of detail. This doc requires a full copy-edit later on, so I will resolve this for now.

File: doc/source/release/2.0.0-notes.rst (hunk 12 above)
Comment (rgommers) [reply doc/source/release/2.0.0-notes.rst:18]: good idea, added something along these lines.

File: doc/source/release/2.0.0-notes.rst [hunk 13]
```diff
@@ -4,6 +4,93 @@
...
+The "no free lunch" theorem dictates that there is a price to pay for all these
+API and behavior improvements and better future extensibility. This price is:
+
+1. Backwards compatibility - clear errors in most cases; see the Deprecation
+   and Compatibility sections below, and the :ref:`numpy-2-migration-guide`
+   (note: there is a ``ruff`` mode to auto-fix many things in Python code)
+2. ABI change - impact on downstream libraries and binaries. See
+   :ref:`numpy-2-abi-handling`.
+   TODO: main messages.
+
```
[the remark this answers is not in the document]
Comment (rgommers) [reply doc/source/release/2.0.0-notes.rst:107]: I would not mention that now. Let's first see how it goes. Numba is usually 3-6 months late. If all of scipy/matplotlib/pandas/scikits have compatible releases up before we do 2.0.0 final, then I don't think we need to say this.

**Comment by rgommers** [comment]
Thanks, accepted all suggestions that are resolved.

**Comment by mattip** [comment]
As expected, #25943 causes pandas import to fail, which [fails the docs](https://app.circleci.com/pipelines/github/numpy/numpy/25449/workflows/896dda60-8d9d-4d90-8c36-cfa3cb2955a8/jobs/38729) build.

**Comment by BvB93** [comment]
> Let me ping a few people for specific categories:
>     * Static typing: @BvB93 (no major changes I think, but want to make sure)

Can confirm, yes; just regular maintenance to reflect the runtime changes.

**Comment by charris** [comment]
close/reopen

**Comment by mattip** [comment]
Close/reopen to get latest dev wheels

**Comment by mattip** [comment]
Maybe we should skip the doctest tests for now, until [a pandas fix](https://github.com/pandas-dev/pandas/pull/57774) for the dtype is merged and new nightly wheels are uploaded?

**Comment by mattip** [comment]
Instead of skipping doctests, we could move them later in the CI steps, which AFAICT would still upload artifacts.

**Comment by rgommers** [comment]
I had the same thought. We use next to nothing from pandas, so it's a little annoying that our doc builds are blocked on it. However, if we break pandas we usually also break matplotlib, so in those cases it won't help.

**Comment by mattip** [comment]
> so in those cases it won't help.

Right, in general it won't help. But for this PR specifically it will: #25962, which was merged to HEAD, does upload artifacts. So rebasing this off `main` will clear the merge conflict and allow seeing the rendered doc (as well as checking for new warnings).

**Comment by rgommers** [comment]
Thanks! Will rebase after the next update - which should make this merge-able.

**Comment by rgommers** [comment]
> The rendered doc is [here](https://output.circle-artifacts.com/output/job/e9991362-1368-47d6-ada1-47f1c98112b5/artifacts/0/doc/build/html/release/2.0.0-notes.html). There some warnings it would be nice to clean up if possible

We really have to get to zero warnings and turn on the "error on warning" behavior in CI. I'd say that at least 75% of all PRs that add release notes are introducing new issues.

**Comment by rgommers** [comment]
I think this is good to go now. It deserves another large copy-edit of the full notes, but that can be done in a few weeks time. This should be good enough for 2.0.0b1.

**Comment by mattip** [comment]
[here](https://output.circle-artifacts.com/output/job/26282d8b-1e60-423e-a58f-5e9712f79ba2/artifacts/0/doc/build/html/release/2.0.0-notes.html) are the rendered docs.

**Comment by mattip** [comment]
I am not sure how to review this, so I will just put it in as-is. Anyway it will not successfully be displayed as the devdocs until the pandas doctest passes and the build is un-broken. It is evident a lot of work went into this, thanks to all who contributed. There are some formatting glitches: some items are lists and some are topics, it would be nice to try to straighten that out for the next version.

**Comment by mattip** [comment]
Thanks @rgommers

**Comment by seberg** [comment]
Thanks!

> the pandas doctest passes and the build is un-broken

Would it be easy to work around?  I am slightly worried that the next in line is matplotlib and that might not be quite as quick.

**Comment by rgommers** [comment]
> I am slightly worried that the next in line is matplotlib and that might not be quite as quick.

Is that in progress / do they need help? I haven't had problems yet with local doc builds, so I think the damage should be limited.

I'm babysitting the SciPy nightlies now, they will hopefully be up within the next hour.

**Comment by mattip** [comment]
> Would it be easy to work around? 

I think a PR to skip the doctest tests and then another to revert it would work, we would merge the second one once it passes.

**Comment by seberg** [comment]
> Is that in progress / do they need help? I haven't had problems yet with local doc builds, so I think the damage should be limited.

OK, cool, for me it is just an unknown for whether they might hit pybind11 issues mainly (I doubt it from my first look, but hard to be sure).

Let's just see, maybe this is all very quick, if not we can still rearrange the upload.

**Comment by charris** [comment]
I'm not worried about the failing doc tests, just ignore them. They don't block the wheel releases and the 2.0.0 release is about a month away, which is when we need to generate the docs.

**Pull request merged**